posts/Stef/2019-11-18-hibernate-orm-with-panache-in-quarkus.adoc

```diff
@@ -0,0 +1,465 @@
+
+- The generated ID field. Very often, all your entities will use the same auto-generate ID type.
+- All those property accessors that do nothing in the entity. They are required for encapsulation, and because Java
+does not support first-class properties in the language. Most people either generate them from their IDE, or using
+https://projectlombok.org/features/GetterSetter[Lombok].
+- All those `persist`/`delete`/`findById`/`findAll` methods on every DAO. All DAOs have them.
+- Those DAO queries all start with `FROM Person` and have to repeat the `Person.class` all over the place.
+
+== A first look at what Hibernate ORM with Panache can do for us
+
+Let's jump forward to https://quarkus.io/[Quarkus] and in particular, what 
+https://quarkus.io/guides/hibernate-orm-panache[Hibernate ORM with Panache] can do for us. It turns out to be quite a lot.
+
+Quarkus allows us to do a lot of bytecode modification at build-time, which (among many benefits) lets us side-step
+Java's lack of support for first-class properties by:
+
+- Writing public fields instead of private+getter+setter
+- Hibernate ORM with Panache will actually generate any getter+setter that you did not write, and
+- It will replace all field accesses with accesses to the getters and setters.
+
+This system allows us to write code as if we were using public fields, but behind the scenes, we still get encapsulation
+and forward-compatibility if we ever add getters or setters that do more than just access the field.
+
+On top of that, Hibernate ORM with Panache comes with support for DAOs that already have a lot of the methods that
+you commonly write.
+
+We can thus rewrite our previous entity class by extending `PanacheEntity` which comes with a predefined auto-generated
+ID field:
+
+[source%nowrap,java]
+----
+@Entity
+public class Person extends PanacheEntity {
+
+    public String firstName;
+    public String lastName;
+    public Date birth;
+}
+----
+
+And we can rewrite our DAO by extending `PanacheRepository` to get all the common methods:
+
+[source%nowrap,java]
+----
+@Singleton
+public class PersonDao implements PanacheRepository<Person>{
+    
+    public List<Person> findByName(String lastName){
+        return find("lastName", lastName).list();
+    }
+    
+    public List<Person> findBornAfter(Date date){
+        return find("birth > :date", Parameters.with("birth", date)).list();
+    }
+}
+----
+
+Note that the `find` convenience method allows HQL, but also _simplified_ HQL (think of it as contextualized HQL):
+
+- If your query is empty, it expands to `FROM <entityType>`
+- If your query starts with `FROM` or `SELECT`, it is left alone as HQL
+- If your query starts with `ORDER BY...` it expands to `FROM <entityType> ORDER BY...`
+- If your query only has a single property and argument, it expands to `FROM <entityType> WHERE <property> = <argument>`
+- Otherwise, your query is taken as a `WHERE...` clause and expands to `FROM <entityType> WHERE...`
```

Comment:
This saves the user a bit of typing, but as a reader I'm thinking of a tradeoff between the benefit here vs. having more "magic" happening by the framework. If a user gets their simplified query wrong, they will need to debug the query going to the DB.  Perhaps we should include a pointer to SQL logging or the `PanacheEntity.find` javadoc which outlines this list?

Reply:
I don't think we publish our javadoc yet, but I'll add a pointer to the guide.